odoo/orm/fields_properties.py

```diff
@@ -105,23 +105,27 @@ def convert_to_cache(self, value, record, validate=True):
         if not value:
             return None
 
-        if isinstance(value, dict):
-            # avoid accidental side effects from shared mutable data
-            return copy.deepcopy(value)
-
-        if isinstance(value, str):
-            value = json.loads(value)
-            if not isinstance(value, dict):
-                raise ValueError(f"Wrong property value {value!r}")
-            return value
+        if not isinstance(value, (str, dict, list)):
+            raise ValueError(f"Wrong property type {type(value)!r}")
 
         if isinstance(value, list):
             # Convert the list with all definitions into a simple dict
             # {name: value} to store the strict minimum on the child
             self._remove_display_name(value)
-            return self._list_to_dict(value)
+            value = self._list_to_dict(value)
+        elif isinstance(value, str):
+            value = json.loads(value)
+            if not isinstance(value, dict):
+                raise ValueError(f"Wrong property value {value!r}")
+        if not validate or not record._ids:
+            return value
+        # writing the same truthy value to properties with different definitions doesn't make sense
+        definition = self._get_properties_definition(record[0])
+        value = self._dict_to_list(value, definition)
+        res_ids_per_model = self._get_res_ids_per_model(record.env, [value], validate=False)
+        self._parse_json_types(value, record.env, res_ids_per_model)
```

Comment:
it means that now `_parse_json_types` is called 2 times, once here and once in `convert_to_read_multi`, is it intended ?

Reply:
> But ORM field read may happen many times even in for loop

The properties field is not meant to be used with `record.field`, before `_parse_json_types` was not called in a such case, do we really need to check the existence of the selection, etc when doing record.field ? it's confusing that we check the existence of some properties, but not others

Reply:
> For business it is required to call _parse_json_types for the result of record.field

Why ? 

> But as a benifit, in the test script, since usually we don't unlink records and you know the records used in the test always exists in the transcation, and the definition is not changed. Developers can easily assert the result using record.field without _parse_json_types.

Yes, but then if they use m2o they will need to use read, it's really not obvious with your change (while it was before, since nothing was checked)

Reply:
> Why ?

because the `record.field` never checks data
`many2one` records might be unlinked after last `write` (exists change)
definition might be changed after last `write` (format change)
we can do nothing to strictly promise the consistency without calling `_parse_json_types`

> while it was before, since nothing was checked

previously it was checked when `_flush`, the `cache_value` is converted to `read_value`(with exists + format check) and then to the `db_value`

Reply:
OK, I thought you took the advantage of `convert_to_read` when `_flush` to clean and validate data on purpose to get cleaner data in database. So I do the partial validation when `write` as a best effort. I remove the the check in the last commit

Reply:
But since the `record.field` is never validated, in the test you cannot directly use the value. See runbot. We have to refactor many tests. Is it OK to you?

Reply:
> But since the record.field is never validated, in the test you cannot directly use the value

It depend on the test I would say, but ok to adapt them if needed

Reply:
I prefer to have a format check here without the exits check
Otherwise when you `record.write({'attributes': {'many2one_field: (1, 'bob')})` the data will be accepted by the ORM. It should be better to warn the user or drop the incorrect format to avoid future error after read

What do you think?

Reply:
both were valid before no ? why is it an issue ?

Reply:
Note: `convert_to_column_update` -> `convert_to_write` -> `convert_to_read` -> `_parse_json_types` with format and exists checks

Before, the if you write
`record.attributes = {'many2one_field: (1, 'bob')}`
the first check happens in the [`field.__set__` before the real `model.write`](https://github.com/odoo/odoo/blob/e17c528d96288170eeec29fbe51da77d9e0c0c19/odoo/orm/fields.py#L1335)
if you write like `record.write({'attributes': {'many2one_field: (1, 'bob')})`
the first check happens in the [`model._flush`](https://github.com/odoo/odoo/blob/e17c528d96288170eeec29fbe51da77d9e0c0c19/odoo/orm/models.py#L6734)

whatever the value with wrong format cannot be written to the database.
The only way to have bad format by using the definition
1. field definition declear it is a char field
2. `record.write({'attributes': {'field_name: 'value'})`
3. change definition and the field becomes a many2one
4. when `convert_to_read` the exist check in `_get_res_ids_per_model` may fail

In this task, maybe we can check here or add and filter in `_get_res_ids_per_model` to drop ids which are not `int`